File: Orbit_ADCS/Matlab-ADCS/ADCS Julia/Determination/QUEST.jl
include("../Rotations/EPRot.jl")
include("../Tools/newton.jl")
include("../types.jl")

function QUEST(vkb::Matrix{Float64}, vkn::Matrix{Float64}, weight::Vector{Float64})
    B::Matrix{Float64} = zeros(3,3)
    for i::Int64 = 1:size(vkb, 2)
        B = B + weight[i]*vkb[:,i]*vkn[:,i]'
    end

    S::Matrix{Float64} = B + B'
    Σ::Float64 = tr(B)
    Z::Vector{Float64} = [B[2,3] - B[3,2], B[3,1] - B[1,3], B[1,2] - B[2,1]]

    K::Matrix{Float64} = [Σ Z'
                          Z S-Σ.*Matrix(I,3,3)]
    
    f(s) = det(K - s[1].*Matrix(I,4,4))
    λ0::Float64 = sum(weight)
    λopt::Float64 = newton(λ0, f)

    qcr = CRP(inv((λopt + Σ)*Matrix(I,3,3) - S)*Z, "B", "N")
    
    return CRP2dcm(qcr)
end

File: Orbit_ADCS/Matlab-ADCS/Determination/QUEST.m
function var = QUEST(vkb, vkn, weight)
    B = zeros(3,3);
    for i = 1:size(vkb, 2)
        B = B + weight(i)*vkb(:,i)*vkn(:,i)';
    end

    S = B + B';
    sig = trace(B);
    Z = [B(2,3) - B(3,2)
         B(3,1) - B(1,3)
         B(1,2) - B(2,1)];

    K = [sig Z'
         Z   S-sig.*eye(3,3)];
    
    f = @(s) det(K - s(1).*eye(4,4));
    lambda0 = sum(weight);
    lambda_opt = newton(lambda0, f, 1000, 1e-14, 1e-8);

    qcr = CRP(inv((lambda_opt + sig)*eye(3,3) - S)*Z, "B", "N");
    
    var = CRP2dcm(qcr);
end